system/lib/emmalloc.c

```diff
@@ -1342,13 +1350,13 @@ size_t emmalloc_dynamic_heap_size()
   size_t dynamicHeapSize = 0;
 
   MALLOC_ACQUIRE();
-  Region *root = listOfAllRegions;
+  RootRegion *root = listOfAllRegions;
   while(root)
   {
-    Region *r = root;
-    uintptr_t blockEndPtr = ((uint32_t*)r)[2];
+    RootRegion *r = root;
+    uintptr_t blockEndPtr = (uintptr_t)root->endPtr;
     dynamicHeapSize += blockEndPtr - (uintptr_t)r;
-    root = ((Region*)((uint32_t*)root)[1]);
+    root = root->next;
```

Comment:
Looks like some glitch also in the original version of this function, and this could read simply

```c
while(root)
{
  dynamicHeapSize += (uintptr_t)root->endPtr - (uintptr_t)root;
  root = root->next;
}
```